### PR TITLE
Add tests to ensure MTE tags are preserved across FFI boundaries

### PR DESCRIPTION
**Pull request by dheaton-arm**
Added run-make tests to verify that, between a Rust-C FFI boundary in both directions, any MTE tags included in a pointer are preserved for the following pointer types, as well as any information stored using TBI:
- int
- float
- string
- function

try-job: aarch64-gnu

### CODE REVIEW

File: tests/run-make/mte-ffi/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Tests that MTE tags and values stored in the top byte of a pointer (TBI) are
+// preserved across FFI boundaries (C <-> Rust).
+// This test does not require MTE: whilst the test will use MTE if available, if it is not,
+// arbitrary tag bits are set using TBI.
+
+//@ only-aarch64
+//@ only-linux
+//@ only-gnu
```
Comment (jieyouxu) [inline tests/run-make/mte-ffi/rmake.rs:8]: Question: this is intended to be only tested for `aarch64-unknown-linux-gnu`?

File: tests/run-make/mte-ffi/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//@ run-pass
```
Comment (jieyouxu) [inline tests/run-make/mte-ffi/rmake.rs:9]: Suggestion: this is not needed, rmake.rs can only be run-pass.

File: tests/run-make/mte-ffi/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        .target(target())
+        .linker("aarch64-linux-gnu-gcc")
+        .run();
+    cc().input(format!("bar_{variant}.c"))
+        .input(dynamic_lib_name("foo"))
+        .out_exe("test")
+        .args(&flags)
+        .run();
+    run("test");
+    println!("\tpassed");
```
Comment (jieyouxu) [inline tests/run-make/mte-ffi/rmake.rs:38]: Nit: not needed, the test status will indicate that it passed.

**Review by jieyouxu — COMMENTED** [review]
Hi, thank you for the PR! I can only advise on the rmake.rs test itself, I'm trying to find you a reviewer who's more familiar with aarch64's MTE behaviors.

File: tests/run-make/mte-ffi/rmake.rs (hunk 1 above)
Comment (dheaton-arm) [reply tests/run-make/mte-ffi/rmake.rs:8]: Yes - I tried `//@ only-aarch64-unknown-linux-gnu` (which I think should work from the compiletest docs, unless I've misunderstood that?) but that kept complaining that it didn't recognise it.

The reason is just that the linker has to be specified when cross-compiling, and I'm not able to easily test other targets.

Comment (jieyouxu) [reply tests/run-make/mte-ffi/rmake.rs:8]: > Yes - I tried `//@ only-aarch64-unknown-linux-gnu` (which I think should work from the compiletest docs, unless I've misunderstood that?) but that kept complaining that it didn't recognise it.

I think you need to add this specifically to a list of known directives (a temporary hack to catch unknown directives while we try to rework compiletest directive parsing) here:

https://github.com/rust-lang/rust/blob/1ddedbaa5919b7b3e70d984660e21e844c615c97/src/tools/compiletest/src/command-list.rs#L4

> The reason is just that the linker has to be specified when cross-compiling, and I'm not able to easily test other targets.

Could you leave that as a comment before/after the ignore? It helps test readers understand *why* it's only the particular target.

**Comment by Amanieu** [comment]
The ARM-specific parts look fine to me. It's just testing that pointer bits are properly preserved over FFI.

Back to you for the rmake part.

r? jieyouxu

**Comment by dheaton-arm** [comment]
Sorry for the delay!

@rustbot review

**Review by jieyouxu — APPROVED** [review]
Thanks!

**Comment by tgross35** [comment]
@bors r- retry

Same as https://github.com/rust-lang/rust/pull/128363, already merged

**Pull request merged**